[PATCH v8] example/l2fwd: add argument to get packet length

ksatha opened the pull request:
Present l2fwd consider packet length 1536 for pool creation, it limits
the packet length to X * 1536 when platform supports X number of segments.
This new command line argument along with pool capabilities will create
packet pool to support packets up to 64K MTU.

Signed-off-by: Satha Rao <skoteshwar@marvell.com>

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -105,6 +108,7 @@ typedef struct {
 	uint64_t vec_tmo_ns;    /* Vector formation timeout in ns */
 	uint32_t vec_size;      /* Vector size */
 	int verbose;		/* Verbose output */
+	int packet_len;		/* Maximum pacekt length supported */
```
Typo: "pacekt".

ksatha replied:
ACK

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -1507,6 +1511,7 @@ static void usage(char *progname)
 	       "  -x, --vec_size <num>    Vector size (default %i).\n"
 	       "  -z, --vec_tmo_ns <ns>   Vector timeout in ns (default %llu ns).\n"
 	       "  -P, --promisc_mode      Enable promiscuous mode.\n"
+	       "  -l, --packet_len <len>  Maximum length of packets supported.\n"
```
Default value could be printed here.

ksatha replied:
ACK

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -2035,13 +2045,23 @@ int main(int argc, char *argv[])
 		printf("\nWarning: packet length reduced to %u\n\n", pkt_len);
 	}
 
+	seg_len = pkt_len > MIN_SEG_LEN ? MIN_SEG_LEN : pkt_len;
+	if (seg_len < (pkt_len / pool_capa.pkt.max_segs_per_pkt))
+		seg_len = pkt_len / pool_capa.pkt.max_segs_per_pkt;
```
Instead of this logic, would it be better to make segment length also a command line option?

ksatha replied:
Adding segment length option will be good suggestion, will submit revised patch with new argument.
And I will move this logic to under no argument case, because if user requested bigger packet length(say 16K) and platform supports limited segments(assume 8), keeping default segment length (1536) will be problem here, so we need to tune segment length based on new packet length.

JereLeppanen reviewed (commented):
Thanks, this looks useful.

The patch subject line prefix should be "test: l2fwd:".

ksatha commented:
> Thanks, this looks useful.
> 
> The patch subject line prefix should be "test: l2fwd:".

Thanks for reviewing. Will change subject in next revision.

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -105,6 +108,8 @@ typedef struct {
 	uint64_t vec_tmo_ns;    /* Vector formation timeout in ns */
 	uint32_t vec_size;      /* Vector size */
 	int verbose;		/* Verbose output */
+	int packet_len;		/* Maximum packet length supported */
+	uint32_t seg_len;	/* Pool segment length */
```
Minor thing, but these should be of the same type, it looks funny if they're not.

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -1507,6 +1512,8 @@ static void usage(char *progname)
 	       "  -x, --vec_size <num>    Vector size (default %i).\n"
 	       "  -z, --vec_tmo_ns <ns>   Vector timeout in ns (default %llu ns).\n"
 	       "  -P, --promisc_mode      Enable promiscuous mode.\n"
+	       "  -l, --packet_len <len>  Maximum length of packets supported(default length 1536).\n"
+	       "  -L, --seg_len <len>     Packet pool segment length(default length 1536).\n"
```
The defaults here should be handled the same way as in the other options.

ksatha replied:
ACK

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -2035,13 +2052,30 @@ int main(int argc, char *argv[])
 
+	seg_len = gbl_args->appl.seg_len;
+
+	/* Check weather we have sufficient segments to support requested packet
```
Typo: "weather" -> "whether".

ksatha replied:
ACK

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -2035,13 +2052,30 @@ int main(int argc, char *argv[])
+
+	/* Check weather we have sufficient segments to support requested packet
+	 * length, if not adjust to bigger segment size */
+	if ((seg_len * pool_capa.pkt.max_segs_per_pkt) < pkt_len)
+		seg_len = pkt_len / pool_capa.pkt.max_segs_per_pkt;
```
It would be clearer to use the same calculation for the condition, i.e. if (seg_len < pkt_len / pool_capa.pkt.max_segs_per_pkt).

ksatha replied:
Will change if condition. Regarding warning at line number 2076 we are printing both requested and configured values for user reference

JereLeppanen reviewed (commented):
I added a few mostly minor comments, please have a look.

Also please remove the merge commit and rebase.

ksatha commented:
> I added a few mostly minor comments, please have a look.
> 
> Also please remove the merge commit and rebase.

Thanks for suggestions. Updated patch pushed with fixes can you please review

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -1571,6 +1580,8 @@ static void parse_args(int argc, char *argv[], appl_args_t *appl_args)
 	appl_args->pool_per_if = 0;
 	appl_args->num_pkt = 0;
+	appl_args->packet_len = POOL_PKT_LEN;
+	appl_args->seg_len = POOL_SEG_LEN;
```
Sorry, I should have realized earlier that segment length should probably default to packet length, if the segment length argument is not given.

ksatha replied:
ACK

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -1507,10 +1512,12 @@ static void usage(char *progname)
 	       "  -x, --vec_size <num>    Vector size (default %i).\n"
 	       "  -z, --vec_tmo_ns <ns>   Vector timeout in ns (default %llu ns).\n"
 	       "  -P, --promisc_mode      Enable promiscuous mode.\n"
+	       "  -l, --packet_len <len>  Maximum length of packets supported(default %d).\n"
+	       "  -L, --seg_len <len>     Packet pool segment length(default %d).\n"
```
One more nitpick here: space before "(default", just like the other defaults, above.

ksatha replied:
ACK

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -1571,6 +1577,8 @@ static void parse_args(int argc, char *argv[], appl_args_t *appl_args)
 	appl_args->pool_per_if = 0;
 	appl_args->num_pkt = 0;
+	appl_args->packet_len = POOL_PKT_LEN;
+	appl_args->seg_len = POOL_PKT_LEN;
```
This makes segment length and packet length default to the same value, which is different from segment length defaulting to packet length. To elaborate, I meant that once we've figured out what packet length is used, by processing command line parameters and checking capabilities, segment length should then default to that resulting packet length.

ksatha replied:
Hi JereLeppanen, Thanks for detailed explanation, will integrate this logic in next version

JereLeppanen commented on test/performance/odp_l2fwd.c:
```diff
@@ -1507,10 +1509,12 @@ static void usage(char *progname)
 	       "  -z, --vec_tmo_ns <ns>   Vector timeout in ns (default %llu ns).\n"
 	       "  -P, --promisc_mode      Enable promiscuous mode.\n"
+	       "  -l, --packet_len <len>  Maximum length of packets supported (default %d).\n"
+	       "  -L, --seg_len <len>     Packet pool segment length (default %d).\n"
```
Now this is no longer the default. Maybe something like "(default equal to packet length)".

ksatha replied:
ACK

JereLeppanen reviewed (approved):
OK, looks good. Thanks for bearing with me.

ksatha commented:
> OK, looks good. Thanks for bearing with me.

Thanks for your review and valuable suggestions.

matiaselo commented:
@ksatha Please add Jere's reviewed-by tag to the commit, so it can be merged.

ksatha commented:
> @ksatha Please add Jere's reviewed-by tag to the commit, so it can be merged.

Thanks, updated commit message.
